drivers/sensor/hx711/hx711.c

```diff
@@ -0,0 +1,562 @@
+	gpio_pin_set(data->sck_gpio, hx711_config.sck_pin, true);
+	k_busy_wait(1);
+
+	/* SCK set LOW */
+	gpio_pin_set(data->sck_gpio, hx711_config.sck_pin, false);
+	k_busy_wait(1);
+
+	/* Return DOUT pin state */
+	return gpio_pin_get(data->dout_gpio, hx711_config.dout_pin);
+}
+
+/**
+ * @brief Read HX711 data. Also sets GAIN for the next cycle.
+ *
+ * @param dev Pointer to the hx711 device structure
+ * @param chan Channel to fetch data for.
+ *             Only SENSOR_CHANNEL_WEIGHT is available.
+ *
+ * @retval 0 on success,
+ * @retval -EACCES error if 600000 usec's are elapsed with no data available.
+ *
+ */
+static int hx711_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	int ret = 0;
+	uint32_t count = 0;
+	int i;
+
+	struct hx711_data *data = dev->data;
+
+	if (data->power != HX711_POWER_ON) {
+		return -EACCES;
+	}
+
+	ret = hx711_wait_until_ready(data, 600000);
+	if (ret != 0) {
+		return ret;
+	}
+
+	/* Clock data out. Optionally disable interrupts */
+#ifdef CONFIG_HX711_DISABLE_INTERRUPTS_WHILE_POLLING
+	uint32_t key = irq_lock();
+#endif
+	for (i = 0; i < 24; i++) {
+		count = count << 1;
+		if (hx711_cycle(data)) {
+			count++;
+		}
+	}
+
+	/* set GAIN for next read */
+	for (i = 0; i < data->gain; i++) {
+		hx711_cycle(data);
+	}
+
+#ifdef CONFIG_HX711_DISABLE_INTERRUPTS_WHILE_POLLING
+	irq_unlock(key);
+#endif
+
+	count ^= 0x800000;
+
+	data->reading = count;
+
+	return ret;
+}
+
+/**
+ * @brief Set HX711 gain.
+ *
+ * @param dev Pointer to the hx711 device structure
+ * @param val sensor_value struct. Only val1 is used.
+ *            valid values are :
+ *                HX711_GAIN_128X (default),
+ *                HX711_GAIN_32X,
+ *                HX711_GAIN_64X,
+ *
+ * @retval 0 on success,
+ * @retval -ENOTSUP error if an invalid GAIN is provided or
+ *         -EACCES if 600000 usec's are elapsed with no data available.
+ *
+ */
+static int hx711_attr_set_gain(const struct device *dev,
+			       const struct sensor_value *val)
+{
+	struct hx711_data *data = dev->data;
+
+	switch (val->val1) {
+	case HX711_GAIN_128X:
+		data->gain = HX711_GAIN_128X;
+		break;
+	case HX711_GAIN_32X:
+		data->gain = HX711_GAIN_32X;
+		break;
+	case HX711_GAIN_64X:
+		data->gain = HX711_GAIN_64X;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+	return hx711_sample_fetch(dev, SENSOR_CHAN_WEIGHT);
+}
+
+/**
+ * @brief Set HX711 rate.
+ *
+ * @param dev Pointer to the hx711 device structure
+ * @param val sensor_value struct. Only val1 is used.
+ *            valid values are :
+ *               HX711_RATE_10HZ (default),
+ *               HX711_RATE_80HZ
+ *
+ * @retval 0 on success,
+ * @retval -ENOTSUP error if an invalid rate value is passed.
+ *
+ */
+static int hx711_attr_set_rate(struct hx711_data *data,
+			       const struct sensor_value *val)
+{
+	int ret;
+
+	switch (val->val1) {
+	case HX711_RATE_10HZ:
+	case HX711_RATE_80HZ:
+		data->rate = val->val1;
+		ret = gpio_pin_set(data->rate_gpio,
```

Comment:
Please check if rate_gpio is not null before call gpio_pin_set(), as 'rate_gpios' is optional into DTS. Example crashes on first call of set_rate() if rate gpio not defined into device tree.

Reply:
Absolutely right :)

Reply:
Resolved by using an added field **cfg->rate_enabled** which is set to 1 if **rate_gpios** can be found in **DT**.